File: go/quantum_vulnerable.go
// Quantum Vulnerable - Broken/Weak Algorithms (DO NOT USE IN PRODUCTION)
package pqc

import (
	"crypto/des"
	"crypto/dsa"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/md5"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha1"

	"golang.org/x/crypto/blowfish"
	"golang.org/x/crypto/ripemd160"
	"golang.org/x/crypto/curve25519"
)

// Broken Hash (VULNERABLE)
func MD5BrokenHash() {
	hash := md5.Sum([]byte("insecure data"))
	_ = hash
}

func MD4BrokenHash() {
	// MD4 is even more broken than MD5
	hasher := md5.New() // Using md5 as placeholder
	hasher.Write([]byte("md4 is broken"))
	_ = hasher.Sum(nil)
}

func SHA1BrokenHash() {
	hash := sha1.Sum([]byte("sha1 collision attacks"))
	_ = hash
}

func RIPEMDWeakHash() {
	hasher := ripemd160.New()
	hasher.Write([]byte("ripemd data"))
	hash := hasher.Sum(nil)
	_ = hash
}

// Weak Symmetric (VULNERABLE)
func DESWeakCipher() {
	key := []byte("8bytkey") // DES uses 8-byte keys
	block, _ := des.NewCipher(key)
	_ = block
}

func TripleDES3DESCipher() {
	key := make([]byte, 24) // 3DES uses 24-byte keys
	block, _ := des.NewTripleDESCipher(key)
	_ = block
}

func BlowfishCipher() {
	key := []byte("secret key")
	cipher, _ := blowfish.NewCipher(key)
	_ = cipher
}

func RC4StreamCipher() {
	// RC4 is completely broken - don't use
	// Using placeholder as rc4 is deprecated
	key := []byte("rc4 key")
	_ = key
}

// Shor Vulnerable - Asymmetric (VULNERABLE to quantum computers)
func RSAEncryption() {
	privateKey, _ := rsa.GenerateKey(rand.Reader, 2048)
	publicKey := &privateKey.PublicKey

	ciphertext, _ := rsa.EncryptPKCS1v15(rand.Reader, publicKey, []byte("data"))
	_ = ciphertext
}

func DSASignature() {
	var params dsa.Parameters
	dsa.GenerateParameters(&params, rand.Reader, dsa.L2048N256)

	var privateKey dsa.PrivateKey
	privateKey.Parameters = params
	dsa.GenerateKey(&privateKey, rand.Reader)
}

func ECDSASignature() {
	privateKey, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	message := []byte("message to sign")
	hash := sha256.Sum256(message)
	r, s, _ := ecdsa.Sign(rand.Reader, privateKey, hash[:])
	_ = r
	_ = s
}

func DiffieHellmanKeyExchange() {
	var privateKey [32]byte
	rand.Read(privateKey[:])
	var publicKey [32]byte
	curve25519.ScalarBaseMult(&publicKey, &privateKey)
}

func ECDHKeyExchange() {
	privateKey, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	// ECDH would use the private key for key agreement
	_ = privateKey.PublicKey
}

// Vulnerable Curves
func Secp256k1Curve() {
	// secp256k1 - used in Bitcoin
	privateKey, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	_ = privateKey
}

func P384Curve() {
	privateKey, _ := ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
	_ = privateKey
}

func P521Curve() {
	privateKey, _ := ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
	_ = privateKey
}

func Curve25519Key() {
	var privateKey [32]byte
	rand.Read(privateKey[:])
	var publicKey [32]byte
	curve25519.ScalarBaseMult(&publicKey, &privateKey)
}

func Ed25519Signature() {
	// Ed25519 - vulnerable to Shor's algorithm
	import "crypto/ed25519"
	publicKey, privateKey, _ := ed25519.GenerateKey(rand.Reader)
	signature := ed25519.Sign(privateKey, []byte("message"))
	_ = publicKey
	_ = signature
}

File: rust/quantum_vulnerable.rs
// Quantum Vulnerable - Broken/Weak Algorithms (DO NOT USE IN PRODUCTION)
use md5::Md5;
use sha1::Sha1;
use des::Des;
use rsa::{RsaPrivateKey, RsaPublicKey, PaddingScheme};
use p256::ecdsa::{SigningKey, VerifyingKey};
use x25519_dalek::{EphemeralSecret, PublicKey};

// Broken Hash (VULNERABLE)
fn md5_broken_hash() {
    let hash = Md5::digest(b"insecure data");
}

fn md4_broken_hash() {
    use md4::Md4;
    let hash = Md4::digest(b"md4 is broken");
}

fn sha1_broken_hash() {
    let hash = Sha1::digest(b"sha1 collision attacks");
}

fn ripemd_weak_hash() {
    use ripemd::Ripemd160;
    let hash = Ripemd160::digest(b"ripemd data");
}

// Weak Symmetric (VULNERABLE)
fn des_weak_cipher() {
    use des::cipher::{BlockEncrypt, KeyInit};
    let cipher = Des::new(&key.into());
}

fn triple_des_3des_cipher() {
    use des::TdesEde3;
    let cipher = TdesEde3::new(&key.into());
}

fn rc4_stream_cipher() {
    // RC4 is completely broken
    let rc4 = Rc4::new(&key);
}

fn blowfish_cipher() {
    use blowfish::Blowfish;
    let cipher = Blowfish::new(&key.into());
}

fn idea_cipher() {
    // IDEA cipher - legacy
    let idea = Idea::new(&key);
}

// Weak MAC (VULNERABLE)
fn hmac_md5_weak() {
    use hmac::Hmac;
    type HmacMd5 = Hmac<Md5>;
    let mut mac = HmacMd5::new_from_slice(key).unwrap();
}

fn hmac_sha1_weak() {
    type HmacSha1 = Hmac<Sha1>;
    let mut mac = HmacSha1::new_from_slice(key).unwrap();
}

// Shor Vulnerable - Asymmetric (VULNERABLE to quantum computers)
fn rsa_encryption() {
    let mut rng = rand::thread_rng();
    let private_key = RsaPrivateKey::new(&mut rng, 2048).unwrap();
    let public_key = RsaPublicKey::from(&private_key);
    let encrypted = public_key.encrypt(&mut rng, PaddingScheme::PKCS1v15Encrypt, &data);
}

fn dsa_signature() {
    use dsa::{SigningKey, VerifyingKey};
    let signing_key = SigningKey::generate(&mut rng);
}

fn ecdsa_p256_signature() {
    let signing_key = SigningKey::random(&mut rng);
    let verifying_key = VerifyingKey::from(&signing_key);
}

fn diffie_hellman_key_exchange() {
    let secret = EphemeralSecret::new(&mut rng);
    let public = PublicKey::from(&secret);
    let shared_secret = secret.diffie_hellman(&their_public);
}

fn ecdh_key_exchange() {
    use p256::ecdh::EphemeralSecret;
    let secret = EphemeralSecret::random(&mut rng);
    let shared = secret.diffie_hellman(&peer_public);
}

fn elgamal_encryption() {
    let elgamal = ElGamal::new(p, g);
    let (c1, c2) = elgamal.encrypt(&public_key, &message);
}

// Vulnerable Curves
fn secp256k1_curve() {
    use k256::ecdsa::SigningKey as Secp256k1Key;
    let key = Secp256k1Key::random(&mut rng);
}

fn p384_curve() {
    use p384::ecdsa::SigningKey as P384Key;
    let key = P384Key::random(&mut rng);
}

fn curve25519_key() {
    use curve25519_dalek::scalar::Scalar;
    let scalar = Scalar::random(&mut rng);
}

fn ed25519_signature() {
    use ed25519_dalek::{Keypair, Signer};
    let keypair = Keypair::generate(&mut rng);
    let signature = keypair.sign(message);
}
